### ISSUE TITLE
Updating the Documentation page

### ISSUE THREAD

**Issue by reubenharry** [haskell-infra/www.haskell.org#189]
This issue is somewhat related to #61 and #136 and the more general goal of making the Haskell "on-boarding" experience smooth.

Below is a suggested new version of the current Documentation page (https://www.haskell.org/documentation/). I don't have any web design skills, so am just presenting it as Markdown, which is not to say that a fancier layout wouldn't look better. The general philosophy is to present fewer options, so that there's a clearer, more centrally endorsed pathway of what to look at, and less chances of broken links. The equivalent Rust page (https://www.rust-lang.org/learn) seems to me like something to aim for (Julia's one - https://julialang.org/learning/ - feels a bit cluttered IMO).

Beyond updating the page, I think it should also be more prominently linked from the front page (maybe just a larger link - see Rust), and perhaps named "Learn", like Rust's, instead of the drier "Documentation".  

I'm also assuming that the Get Started instructions would be elsewhere (i.e. a link to GHCUP on the front page).

This is intended as a draft, not a complete version. The idea is to get a sense of whether people think this is
a) generally a good idea, but disagree with the specifics
b) bad idea, makes things worse, but thanks for the suggestions!
c) good idea, and here are some further suggestions

-------

# Resources for Newcomers 

New to Haskell and want to learn the basics of the world's leading pure, lazy and statically typed language? You're in the right place!

## Learning the language

[Learn You a Haskell](http://learnyouahaskell.com/) is a community favorite book. There's also a [notebook version](https://github.com/jamesdbrock/learn-you-a-haskell-notebook) if you like to learn interactively.

<details>
  <summary>Other intro Haskell books (click to expand)</summary>
  
*   \[$$\] [Haskell from the Very Beginning](https://www.haskellfromtheverybeginning.com/)
*   [Learn You a Haskell for Great Good!](http://learnyouahaskell.com/)
*   [Real World Haskell](http://book.realworldhaskell.org/)
*   \[$$\] [Haskell Programming from first principles](http://haskellbook.com)
*   \[$$\] [Beginning Haskell](http://www.apress.com/9781430262503)
*   \[$$\] [Thinking Functionally with Haskell](http://www.cambridge.org/us/academic/subjects/computer-science/programming-languages-and-applied-logic/thinking-functionally-haskell)
*   \[$$\] [Programming in Haskell](http://www.cs.nott.ac.uk/~pszgmh/pih.html)
*   \[$$\] [Haskell: The Craft of Functional Programming](http://www.haskellcraft.com/craft3e/Home.html)
*   \[$$\] [The Haskell School of Music](http://euterpea.com/haskell-school-of-music/)
*   \[$$\] [Get Programming with Haskell](https://www.manning.com/books/get-programming-with-haskell)
*   [Learn Haskell by building a blog generator](https://lhbg-book.link)
*   \[$$\] [Effective Haskell (beta)](https://www.pragprog.com/titles/rshaskell/effective-haskell/)

</details>

[The online course from UPenn](https://www.seas.upenn.edu/~cis194/spring13/lectures.html). Thorough and clear, and the homework is optional :)

<details>
  <summary>Other courses and tutorials (click to expand)</summary>
  
*   [University of Pennsylvania's CIS 194](https://www.seas.upenn.edu/~cis194/)
*   [Data61 Functional Programming Course](https://github.com/data61/fp-course)
*   [University of Virginia's CS 1501](http://shuklan.com/haskell/)
*   [Stanford's cs240h](http://www.scs.stanford.edu/14sp-cs240h/)
*   [Hendrix's CSCI 360](http://ozark.hendrix.edu/~yorgey/360/f16/)
*   [A Gentle Introduction to Haskell](https://www.haskell.org/tutorial/)
*   [Happy Learn Haskell Tutorial](https://www.happylearnhaskelltutorial.com/)
*   [Yet Another Haskell Tutorial](http://en.wikibooks.org/wiki/Haskell/YAHT/Preamble)
*   [Write Yourself a Scheme in 48 Hours](http://en.wikibooks.org/wiki/Write_Yourself_a_Scheme_in_48_Hours)
*   [Write Yourself a Scheme 2.0](https://wespiser.com/writings/wyas/home.html)
*   [Learning Haskell](http://learn.hfm.io)
</details>

## How to use Haskell libraries

A single link to a nice explanation of Hackage, Haddock (focusing on how to read docs) [I have yet to write this, or find a good resource to link to]
A single link to an explanation of Hoogle, with a nice example (and perhaps with a Gif of the VSCode hoogle extension searching for something cool)

# Resources for Advanced Haskellers

Know the basics, but unsure how to write real-world Haskell code? Here are some key resources.

## What are the features I haven't learned about?

[A friendly survey of important language concepts and features](http://dev.stephendiehl.com/hask/)

[The 2010 Haskell language report](https://www.haskell.org/onlinereport/haskell2010/)

## What libraries should I use?

[Overview of the package ecosystem](https://github.com/Gabriel439/post-rfc/blob/main/sotu.md)
[Haskell (Almost) Standard Libraries](https://leanpub.com/haskell-stdlibs)

## Library tutorials

[Parsing](https://markkarpov.com/tutorial/megaparsec.html)
[Parallel and Concurrent Haskell](https://www.oreilly.com/library/view/parallel-and-concurrent/9781449335939/)
[Developing Web Applications with Haskell and Yesod](https://www.oreilly.com/library/view/developing-web-applications/9781449336868/)

## Getting to know the compiler

The [documentation](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/) to the main Haskell compiler, GHC

---

**Comment by tomjaguarpaw** [comment]
This is great stuff, thank you for putting it together. I will come back and help flesh out these ideas at some point in the near future.

**Comment by soupi** [comment]
Some more comments:

I think it's important to recognize that there are many different kinds of people who might be interested in Haskell, such as:

1. Academics interested in math and research
2. Curious programmers interested in expanding their horizons
3. Industrial programmers interested in adding another productive tool to their toolbox
4. Programmers interested in intellectual challenges

...and so on. I don't think there's no one size fits all here - these groups likely want different things from Haskell and a getting started guide that is tailored to one group will probably make it harder for the other groups.

For example, I think that the industrial programmer will benefit from reading [cs240h basics](https://www.scs.stanford.edu/16wi-cs240h/slides/basics.html) more than from reading LYAH, and the academic might benefit more from reading the Haskell report. Most of them will probably be interested in a super quick tour to what Haskell looks like.

Another example: should the getting started page mention category theory? For the math enthusiasts and academics, probably yes! For the industrial programmers? There's a common conception that to use Haskell you must learn category theory. Maybe this will enforce the wrong impression that category theory is a must which will drive some people away?

In short, the target audience for the getting started page should be clear. Might even be worth considering making multiple getting started pages depending on the audience you'd like to cater to.

---

> Resources for Advanced Haskellers
> Know the basics, but unsure how to write real-world Haskell code? Here are some key resources.

Do note that not being able to use Haskell to build things is not a sign of an advanced Haskeller, but of a beginner. A first learning resource should take care of that (even if some of them don't...). The way I see it an advanced Haskell is one who can already build stuff with Haskell, but is interested in getting more out of it with more specific things like reducing boilerplate, improving correctness/static guarantees, improving performance, wrapping C libraries, etc.

> What are the features I haven't learned about?
> [A friendly survey of important language concepts and features](http://dev.stephendiehl.com/hask/)

I think WIWIK can be overwhelming rather then friendly. It is a very useful piece, especially for learning specific things or searching for an answer to a question, but I think this phrasing presets it as an article to read.

> Library tutorials

This is kind of related to [fp complete's haskell page](https://www.fpcomplete.com/haskell/learn/). I think it's a good idea in general, though maybe in a separate page. I also think that a section of beginner friendly project-based tutorials (for example a website, a game, a compiler, etc) can be very useful. I would also mention [the haskell phrasebook](https://typeclasses.com/phrasebook).

**Comment by reubenharry** [comment]
Thanks, this is really useful feedback.

The point about the different groups coming to Haskell is well-taken, but here's my thinking. I'm less worried about groups 1 and 4: academics will be more familiar with the motivations for Haskell and less concerned about its differences to conventional languages used in industry. Programmers interested in intellectual challenges will probably be willing to do their own research into resources too. It's groups 2 and 3 that I think a learning page on haskell.org should principally target, basically to lure them in (or at least avoid repelling them).

That's true re. LYAH. Maybe that detail could be changed to cs240h or similar, or one of the many other Haskell intro books. Or LYAH could be offered as a resource for people with less experience programming.

Maybe "Advanced" could be substituted (copying the Rust page here) for "Grow with Haskell" and the actually advanced features could be "Master Haskell". Re WIWIK, that's a great point too.

I hadn't seen the Haskell phrasebook, it's nice too. One thing I'm aware of is that there are many many good resources, but my thinking is that it's best to pick a small subset, so that subsequent attention can be given to making sure those are as good as possible and kept up to date (again see the Rust page). I'm envisioning a virtuous cycle where a small set of great resources get maintained and improved over time.

**Comment by hasufell** [comment]
> > [Learn You a Haskell](http://learnyouahaskell.com/) is a community favorite book.
> 
> Consider this: LYAH has been the most popular Haskell book in the past 10 years, and Haskell is still known to be very difficult to learn, so maybe putting LYAH front and center isn't exactly going to help the Haskell on-boarding experience?

LYAH has not been recommended by us IRC users for a long time, mainly because it doesn't have any exercises and is not considered an in-depth resource. So I'm strongly -1 on that specific idea.

----

Also note that ghcup now has a "getting started" page with links to proper learning materials: https://www.haskell.org/ghcup/steps/

**Comment by tomjaguarpaw** [comment]
Thanks again to @reubenharry for putting this together and @soupi for @hasufell for the comments. The point about LYAH is pretty compelling to me. Personally I've never read it, but I understand what @soupi means: if so many people have tried it, yet Haskell still has this reputation as "hard to learn", then can it really be such a great introduction? There must be better materials for explaining Haskell. [@bitemyapp's "How should I learn Haskell guide"](https://github.com/bitemyapp/learnhaskell#how-should-i-learn-haskell) is tried and tested, I think.

I personally don't have time to drive this forward or even participate much, but I think that converting the "Documentation" page to be somewhat more opinionated is a good idea. It will take time to do work out the exact contents and do the necessary consensus-building in the community, but it will be worthwhile I think.

**Comment by tomjaguarpaw** [comment]
I would advise @reubenharry to keep going, keep iterating and responding to feedback, and hopefully the result will be something will be a nicely polished page, or introduction to Haskell learning resources, and we can look at incorporating it on the site.

**Comment by reubenharry** [comment]
Sounds good. I've started updating the above markdown draft, and will continue. I left LYAH in for now, because it is the only free intro book I'm aware of (Real World Haskell being somewhat out of date) and I at least can attest to having learned the basics of Haskell from it, though I agree that it is flawed.

FP Complete has a bunch of really useful materials, but I assume there are political issues with incorporating those materials with haskell.org? Hopefully not

**Comment by hasufell** [comment]
> Sounds good. I've started updating the above markdown draft, and will continue. I left LYAH in for now, because it is the only free intro book I'm aware of (Real World Haskell being somewhat out of date) and I at least can attest to having learned the basics of Haskell from it, though I agree that it is flawed.
> 

https://github.com/haskell-beginners-2022/course-plan

https://www.cis.upenn.edu/~cis194/spring13/

As linked by https://www.haskell.org/ghcup/steps/#how-to-learn-haskell-proper

These are real learning materials.

**Comment by reubenharry** [comment]
OK, I removed LYAH, by way of compromise :) And since the rest of the books were either not for beginners or not free, I took them out too.

(What I removed is below)

---

[Learn You a Haskell](http://learnyouahaskell.com/) is a friendly place to start if you're new to programming. There's also a [notebook version](https://github.com/jamesdbrock/learn-you-a-haskell-notebook) if you like to learn interactively.

<details>
  <summary>Other intro Haskell books (click to expand)</summary>
 
*  [Parallel and Concurrent Haskell](https://www.oreilly.com/library/view/parallel-and-concurrent/9781449335939/)
*  [Developing Web Applications with Haskell and Yesod](https://www.oreilly.com/library/view/developing-web-applications/9781449336868/)
*   \[$$\] [Haskell from the Very Beginning](https://www.haskellfromtheverybeginning.com/)
*   [Real World Haskell](http://book.realworldhaskell.org/)
*   \[$$\] [Haskell Programming from first principles](http://haskellbook.com)
*   \[$$\] [Beginning Haskell](http://www.apress.com/9781430262503)
*   \[$$\] [Thinking Functionally with Haskell](http://www.cambridge.org/us/academic/subjects/computer-science/programming-languages-and-applied-logic/thinking-functionally-haskell)
*   \[$$\] [Programming in Haskell](http://www.cs.nott.ac.uk/~pszgmh/pih.html)

</details>

**Comment by gbaz** [comment]
I think it is wrong to remove the books entirely from the documentation page. Even though some of the books are paid, they are invaluable resources and they should be listed. Additionally, it is worthwhile to list even the intermediate ones, if they are marked as such. I'm fine with not steering users to them as a first stop, but to remove them altogether will have a bad effect.

I also think the haskell wikibook is fine to refer people to.

In general, the material which is on the page and is removed in this PR (not just books) is good material and worthwhile to keep around. Highlighting a "getting started" section that has a few recommended tutorials, and is opinionated and steers people is a good thing. I like the idea behind it!

However, below that section, or on another page, a more comprehensive list of resources would also be very worthwhile. If people found it more difficult to find e.g. the cabal users guide or the typeclassopedia, or marlow's wonderful book on concurrency, or bird's "thinking functionally in haskell" etc., then that would be a real shame.

**Comment by gbaz** [comment]
A general proposal would be that this material go under "Learning Haskell" at the top of the page (and the links not just say "this documentation" or "this article" but name the cited resource) but below it there's an "Additional Documentation" section that keeps all the existing resources.

The fact that there's a bunch of actual paid print books about Haskell is a good thing -- some people want them, and it speaks to the quality and scope of the resources available. If that information is not easy to come by, it substantially undersells the language and the material available.

**Comment by tomjaguarpaw** [comment]
> I think it is wrong to remove the books entirely from the documentation page

I agree, and I think that probably the smoothest approach would be to create an entirely new page called something like "Learn Haskell" or "Start your Haskell journey" and keep the Documentation page as it is.

**Comment by hasufell** [comment]
> > I think it is wrong to remove the books entirely from the documentation page
> 
> I agree, and I think that probably the smoothest approach would be to create an entirely new page called something like "Learn Haskell" or "Start your Haskell journey" and keep the Documentation page as it is.

It's not really **documentation**. It's a dump of links. That may be useful, but is not what anyone expects of documentation.

**Comment by reubenharry** [comment]
@gbaz I agree that many of the books I removed are very valuable resources that people should definitely have access to. I should emphasize that what I'm advocating is basically what @tomjaguarpaw is saying, which is to make a "Learn Haskell" page with a highly directed set of resources, rather than exposing people to all good resources out there.

Having a larger and more complete listing of all good resources is also nice, but as @hasufell says, that isn't really what the current Documentation page is doing. In fact, doing that well seems like a much larger task, where a large subset of the really high quality resources (books and online articles) are carefully sorted and presented.

What I'm proposing is more of a short term easy improvement, which is to take the current documentation page (that I think is too unstructured) and turn it into a "Learn Haskell" page that serves to orient newcomers. I think that would be a net improvement to the Haskell website.

As a compromise, I could add the books as *Further Resources* at the bottom though.

**Comment by gbaz** [comment]
Removing a bunch of useful stuff and not putting it somewhere else is not an improvement. I support creating an upfront section with a "learn haskell" focus. I oppose removing links to resources that don't fit there. They should either be kept on the same page as a lower section, or moved to another easily accessible page.

**Comment by tomjaguarpaw** [comment]
> if links are broken (which as far as I know, no-one is in charge of checking) it gives a really bad impression of the Haskell page

All links are checked, both external and internal, on every CI run:

https://github.com/haskell-infra/www.haskell.org/blob/089dd671ec94bc41d396ae0235bee0dc0e8ec284/.github/workflows/main.yml#L35-L42

**Comment by reubenharry** [comment]
OK, after a bit of thinking, I think you're right that what I'm proposing shouldn't be in replacement of the Documentation page, but in addition. I think what I really want is a "Getting Started" page that recommends a toolchain,  links to a single tutorial, and explains the library system (hackage and hoogle, and an overview of which libraries are for what).

(As well as that, I think the current frontpage of haskell.org should have more propaganda on why to use Haskell, but that's a separate issue.)

Separately, I think the current documentation page isn't adding a huge amount of value in its current state. I think having a page with lots of links is useful, but actually not as useful as it might seem, because people are comfortable with using google. For example, anyone who wants a Haskell book can google "Haskell books" and pick one that looks good to them. I'd prefer to see slightly fewer links, with more explanation of each link. In that vein, a link to Hackage and Stackage with no context is not that useful if you don't know how to use them. But improving this is a lot of hard work.

I'll step back and think how I can update the above proposal to make it a "Get started" page, and not a replacement for the Documentation page.
